builtin/credential/aws/path_login.go

```diff
@@ -888,6 +1030,392 @@ func (b *backend) pathLoginRenew(
+				"auth_type":          "iam",
+				"role_tag_max_ttl":   rTagMaxTTL.String(),
+				"inferredEntityType": inferredEntityType,
+				"inferredEntityId":   inferredEntityId,
+				"account_id":         accountID,
+			},
+			InternalData: map[string]interface{}{
+				"role_name": roleName,
+			},
+			DisplayName: principalName,
+			LeaseOptions: logical.LeaseOptions{
+				Renewable: true,
+				TTL:       roleEntry.TTL,
+			},
+		},
+	}
+
+	if roleEntry.Period > time.Duration(0) {
+		resp.Auth.TTL = roleEntry.Period
+	} else {
+		shortestTTL := b.System().DefaultLeaseTTL()
+		if roleEntry.TTL > time.Duration(0) && roleEntry.TTL < shortestTTL {
+			shortestTTL = roleEntry.TTL
+		}
+
+		maxTTL := b.System().MaxLeaseTTL()
+		if roleEntry.MaxTTL > time.Duration(0) && roleEntry.MaxTTL < maxTTL {
+			maxTTL = roleEntry.MaxTTL
+		}
+		if rTagMaxTTL > time.Duration(0) && rTagMaxTTL < maxTTL {
+			maxTTL = rTagMaxTTL
+		}
+
+		if shortestTTL > maxTTL {
+			resp.AddWarning(fmt.Sprintf("Effective TTL of %q exceeded the effective max_ttl of %q; TTL value is capped accordingly", (shortestTTL / time.Second).String(), (maxTTL / time.Second).String()))
+			shortestTTL = maxTTL
+		}
+
+		resp.Auth.TTL = shortestTTL
+	}
+
+	return resp, nil
+}
+
+// These two methods (hasValuesFor*) return two bools
+// The first is a hasAll, that is, does the request have all the values
+// necessary for this auth method
+// The second is a hasAny, that is, does the request have any of the fields
+// exclusive to this auth method
+func hasValuesForEc2Auth(data *framework.FieldData) (bool, bool) {
+	_, hasPkcs7 := data.GetOk("pkcs7")
+	_, hasIdentity := data.GetOk("identity")
+	_, hasSignature := data.GetOk("signature")
+	return (hasPkcs7 || (hasIdentity && hasSignature)), (hasPkcs7 || hasIdentity || hasSignature)
+}
+
+func hasValuesForIamAuth(data *framework.FieldData) (bool, bool) {
+	_, hasRequestMethod := data.GetOk("request_method")
+	_, hasRequestUrl := data.GetOk("request_url")
+	_, hasRequestBody := data.GetOk("request_body")
+	_, hasRequestHeaders := data.GetOk("request_headers")
+	return (hasRequestMethod && hasRequestUrl && hasRequestBody && hasRequestHeaders),
+		(hasRequestMethod || hasRequestUrl || hasRequestBody || hasRequestHeaders)
+}
+
+func parseIamArn(iamArn string) (string, string, string, error) {
+	fullParts := strings.Split(iamArn, ":")
+	principalFullName := fullParts[5]
+	parts := strings.Split(principalFullName, "/")
+	principalName := parts[1]
+	transformedArn := iamArn
+	sessionName := ""
+	if parts[0] == "assumed-role" {
+		transformedArn = fmt.Sprintf("arn:aws:iam::%s:role/%s", fullParts[4], principalName)
+		sessionName = parts[2]
+	} else if parts[0] != "user" {
+		return "", "", "", fmt.Errorf("unrecognized principal type: '%s'", parts[0])
+	}
+	return transformedArn, principalName, sessionName, nil
+}
+
+func ensureVaultHeaderValue(headers http.Header, requestUrl *url.URL, requiredHeaderValue string) (bool, string) {
```

Comment:
We can do away with the boolean return variable and have this method just return an `error`A `nil` value can indicate that the validation was successful.